downloads/swan-lake-release-notes/swan-lake-2201.5.0/RELEASE_NOTE.md

```diff
@@ -0,0 +1,498 @@
+
+- Added the following new arguments to the `persist init` commands.
+    - `--datastore` - This is used to indicate the preferred database client. Currently, only `mysql` is supported.
+    - `--module` - This is used to indicate the module in which the files are generated.
+
+    For example, 
+    ```bash
+    $ bal persist init --datastore mysql --module db
+    ```
+- Changed the `persist init` command to create a `persist` directory in the Ballerina project and generate a new definition file (`model.bal`) in the `persist` directory if the file does not exist.
+- Restricted to have only one persist model definition per Ballerina package.
+- Removed the `persist push` command support and generate SQL script file in the `persist generate` command. The generated SQL script file needs to be executed manually to create the database table.
+- Changed the `persist generate` command to generate all the Ballerina client, types, `db_config` files, and the SQL script files in the `generated/<module_name>` directory.
+- Renamed the generated files to the following.
+  - `generated_client.bal` -> `persist_client.bal`
+  - `generated_types.bal` -> `persist_types.bal`
+  - `database_configuration.bal` -> `persist_db_config.bal`
+  - `<schema_name>_db_script.sql` -> `script.sql`
+- Changed the code in the generated Ballerina client and types files to support the new changes.
+
+### Improvements
+
+#### Bindgen Tool 
+
+- Added the `--with-optional-types`, `--with-optional-types-param`, and `--with-optional-types-return` command options to the `bal bindgen` command, to support handling Java null values via generated bindings. These command options will generate optional (i.e., nilable) types 
+  for parameter or/and return types in generated Ballerina binding functions.
+
+    Consider the below Java methods and the corresponding Ballerina binding functions generated with and without the new command options.
+    ```java
+    // Parameters and return types having inbuilt object types (`java.lang.String`).
+    public String f1(String str) {
+        return str;
+    }
+    
+    // Parameter and return types having external object arrays (`Foo[]`).
+    public Foo[] f2(Foo[] fooArray) {
+        return fooArray;
+    }
+    ```
+    
+    **Without optional types** 
+    
+    ```bash
+    $ bal bindgen
+    ```
+    
+    ```ballerina
+    public function f1(string arg0) returns string {
+        return java:toString(Foo_strParamReturns(self.jObj, java:fromString(arg0))) ?: "";
+    }
+
+    public function f2(Foo[] arg0) returns Foo[]|error {
+        handle externalObj = Foo_objArrayParamReturns(self.jObj, check jarrays:toHandle(arg0, "Foo"));
+        Foo[] newObj = [];
+        handle[] anyObj = <handle[]>check jarrays:fromHandle(externalObj, "handle");
+        int count = anyObj.length();
+        foreach int i in 0 ... count - 1 {
+            Foo element = new (anyObj[i]);
+            newObj[i] = element;
+        }
+        return newObj;
```

Comment:
Is this generated code? Can be improved IMO (e.g., query expressions, ..<, etc.).

Reply:
@MaryamZi yes this is generated code from the bindgen tool

Reply:
Ack. Maybe we can look into this separately to identify improvements.

Reply:
https://github.com/ballerina-platform/ballerina-lang/issues/40222